File: Source/AlphaTab/Rendering/Glyphs/BarNumberGlyph.cs
using AlphaTab.Platform;

namespace AlphaTab.Rendering.Glyphs
{
    internal class BarNumberGlyph : Glyph
    {
        private readonly int _number;

        public BarNumberGlyph(float x, float y, int number)
            : base(x, y)
        {
            _number = number;
        }

        public override void DoLayout()
        {
            Renderer.ScoreRenderer.Canvas.Font = Renderer.Resources.BarNumberFont;
            Width = Renderer.ScoreRenderer.Canvas.MeasureText(_number.ToString()) + 5 * Scale;
        }

        public override void Paint(float cx, float cy, ICanvas canvas)
        {
            if (!Renderer.Staff.IsFirstInAccolade)
            {
                return;
            }

            var res = Renderer.Resources;
            var c = canvas.Color;
            canvas.Color = res.BarNumberColor;
            canvas.Font = res.BarNumberFont;

            canvas.FillText(_number.ToString(), cx + X, cy + Y);
            canvas.Color = c;
        }
    }
}

File: packages/alphatab/src/rendering/glyphs/BarNumberGlyph.ts
import { TextBaseline, type ICanvas } from '@coderline/alphatab/platform/ICanvas';
import type { RenderingResources } from '@coderline/alphatab/RenderingResources';
import { Glyph } from '@coderline/alphatab/rendering/glyphs/Glyph';
import type { LineBarRenderer } from '@coderline/alphatab/rendering/LineBarRenderer';
import { ElementStyleHelper } from '@coderline/alphatab/rendering/utils/ElementStyleHelper';
import { NotationElement } from '@coderline/alphatab/NotationSettings';

/**
 * @internal
 */
export class BarNumberGlyph extends Glyph {
    private _number: string;

    public constructor(x: number, y: number, num: number) {
        super(x, y);
        this._number = `${num}  `;
    }

    public override doLayout(): void {
        // TODO: activate this and update paddings accordingly.
        // if (!this.renderer.staff!.isFirstInSystem) {
        //     this.width = 0;
        //     this.height = 0;
        //     return;
        // }

        this.renderer.scoreRenderer.canvas!.font = this.renderer.resources.elementFonts.get(NotationElement.BarNumber)!;
        const size = this.renderer.scoreRenderer.canvas!.measureText(this._number);
        this.width = size.width;
        this.height = size.height;
        this.y -= this.height;
    }

    public override paint(cx: number, cy: number, canvas: ICanvas): void {
        if (!this.renderer.staff!.isFirstInSystem) {
            return;
        }

        using _ = ElementStyleHelper.bar(
            canvas,
            (this.renderer as LineBarRenderer).barNumberBarSubElement,
            this.renderer.bar,
            true
        );

        const res: RenderingResources = this.renderer.resources;
        const baseline = canvas.textBaseline;
        canvas.font = res.elementFonts.get(NotationElement.BarNumber)!;
        canvas.textBaseline = TextBaseline.Top;
        canvas.fillText(this._number, cx + this.x, cy + this.y);
        canvas.textBaseline = baseline;
    }
}
